src/parquet/column/column-writer-test.cc

```diff
@@ -154,8 +166,54 @@ class TestPrimitiveWriter : public PrimitiveTypedTest<TestType> {
   std::unique_ptr<ColumnChunkMetaDataBuilder> metadata_;
   std::unique_ptr<InMemoryOutputStream> sink_;
   std::shared_ptr<WriterProperties> writer_properties_;
+  std::vector<std::shared_ptr<std::vector<uint8_t>>> data_buffer_;
 };
 
+template <typename TestType>
+void TestPrimitiveWriter<TestType>::ReadColumnFully(Compression::type compression) {
+  BuildReader(compression);
+  values_read_ = 0;
+  while (values_read_ < static_cast<int64_t>(this->values_out_.size())) {
+    int64_t values_read_recently = 0;
+    reader_->ReadBatch(this->values_out_.size() - values_read_,
+        definition_levels_out_.data() + values_read_,
+        repetition_levels_out_.data() + values_read_,
+        this->values_out_ptr_ + values_read_, &values_read_recently);
+    values_read_ += values_read_recently;
+  }
+  this->SyncValuesOut();
+}
+
+template <>
+void TestPrimitiveWriter<FLBAType>::ReadColumnFully(Compression::type compression) {
+  BuildReader(compression);
+  this->data_buffer_.clear();
+
+  values_read_ = 0;
+  while (values_read_ < static_cast<int64_t>(this->values_out_.size())) {
+    int64_t values_read_recently = 0;
+    reader_->ReadBatch(this->values_out_.size() - values_read_,
+        definition_levels_out_.data() + values_read_,
+        repetition_levels_out_.data() + values_read_,
+        this->values_out_ptr_ + values_read_, &values_read_recently);
+
+    // Copy contents of the pointers
+    auto data = std::make_shared<std::vector<uint8_t>>(
+        values_read_recently * this->descr_->type_length());
+    uint8_t* data_ptr = data->data();
+    for (int64_t i = 0; i < values_read_recently; i++) {
+      memcpy(data_ptr + this->descr_->type_length() * i,
+          this->values_out_[i + values_read_].ptr, this->descr_->type_length());
+      this->values_out_[i + values_read_].ptr =
+          data_ptr + this->descr_->type_length() * i;
+    }
+    data_buffer_.push_back(data);
```

Comment:
Use emplace_back to move instead of copying, then you can avoid the `shared_ptr`